e2e/clone_test:  Refactor closures to regular functions

RamLavi opened the pull request:
### What this PR does
This PR is refactoring clusre (lambda) functions to regular functions.
Most of the functions only require the virtctl input param, that can be accessed via kubevirt library anyways.
Moving it to regular function simplifies the code.

<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->
Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

RamLavi commented:
/kind cleanup

0xFelix commented on tests/clone_test.go:
```diff
@@ -874,3 +729,148 @@ func createSourceVM(options ...libvmi.Option) (*virtv1.VirtualMachine, error) {
+		},
+	}
+
+	snapshot, err = kubevirt.Client().VirtualMachineSnapshot(snapshot.Namespace).Create(context.Background(), snapshot, v1.CreateOptions{})
```
```suggestion
	snapshot, err := kubevirt.Client().VirtualMachineSnapshot(snapshot.Namespace).Create(context.Background(), snapshot, v1.CreateOptions{})
```

RamLavi replied:
DONE

0xFelix commented on tests/clone_test.go:
```diff
@@ -874,3 +729,148 @@ func createSourceVM(options ...libvmi.Option) (*virtv1.VirtualMachine, error) {
 }
+
+func createSnapshot(vm *virtv1.VirtualMachine) *snapshotv1.VirtualMachineSnapshot {
+	var err error
+
```
```suggestion
```

RamLavi replied:
DONE

0xFelix commented on tests/clone_test.go:
```diff
@@ -874,3 +729,148 @@ func createSourceVM(options ...libvmi.Option) (*virtv1.VirtualMachine, error) {
+	return snapshot
+}
+
+func waitSnapshotReady(snapshot *snapshotv1.VirtualMachineSnapshot) *snapshotv1.VirtualMachineSnapshot {
```
Can you return a new ptr instead of overwriting the one that was passed in?

0xFelix replied:
Possibly only pass in namespace and name of the snapshot?

RamLavi replied:
DONE

0xFelix commented on tests/clone_test.go:
```diff
@@ -874,3 +729,148 @@ func createSourceVM(options ...libvmi.Option) (*virtv1.VirtualMachine, error) {
 	return virtClient.VirtualMachine(vm.Namespace).Create(context.Background(), vm, v1.CreateOptions{})
 }
+
+func createSnapshot(vm *virtv1.VirtualMachine) *snapshotv1.VirtualMachineSnapshot {
```
Consider passing vm.Name and vm.Namespace instead of the full VM if it is not required otherwise.

RamLavi replied:
DONE

0xFelix commented on tests/clone_test.go:
```diff
@@ -874,3 +729,148 @@ func createSourceVM(options ...libvmi.Option) (*virtv1.VirtualMachine, error) {
+	return snapshot
+}
+
+func waitSnapshotContentsExist(snapshot *snapshotv1.VirtualMachineSnapshot) *snapshotv1.VirtualMachineSnapshot {
```
Same

RamLavi replied:
DONE

0xFelix commented on tests/clone_test.go:
```diff
@@ -874,3 +729,148 @@ func createSourceVM(options ...libvmi.Option) (*virtv1.VirtualMachine, error) {
+	return snapshot
+}
+
+func generateCloneFromVMWithParams(sourceVM *virtv1.VirtualMachine, targetVMName string) *clone.VirtualMachineClone {
```
Same

RamLavi replied:
DONE

0xFelix reviewed (commented):
I'd prefer if you could narrow down the values that are passed into the helpers. Instead of ptrs to full objects I'd like to see namespace and name of objects passed in, if nothing else is required.

RamLavi reviewed (commented):
@felix [addressed](https://github.com/kubevirt/kubevirt/compare/ee9d807f43a5b6baad716cb9fed8562aa23fff0d..36e297b94fb39df0fa9daca0206386d667b4785a) review

RamLavi commented:
[Change](https://github.com/kubevirt/kubevirt/compare/ee9d807f43a5b6baad716cb9fed8562aa23fff0d..36e297b94fb39df0fa9daca0206386d667b4785a): addressed reviews

0xFelix commented on tests/clone_test.go:
```diff
@@ -839,22 +694,22 @@ var _ = Describe("VirtualMachineClone Tests", Serial, func() {
-func startCloneVM(virtClient kubecli.KubevirtClient, vm *virtv1.VirtualMachine) (*virtv1.VirtualMachine, error) {
-	patch, err := patch.New(patch.WithAdd("/spec/runStrategy", virtv1.RunStrategyAlways)).GeneratePayload()
+func startCloneVM(vmName, vmNamespace string) (*virtv1.VirtualMachine, error) {
+	patchBytes, err := patch.New(patch.WithAdd("/spec/runStrategy", virtv1.RunStrategyAlways)).GeneratePayload()
 	if err != nil {
 		return nil, err
 	}
 
-	return virtClient.VirtualMachine(vm.Namespace).Patch(context.Background(), vm.Name, types.JSONPatchType, patch, v1.PatchOptions{})
+	return kubevirt.Client().VirtualMachine(vmNamespace).Patch(context.Background(), vmName, types.JSONPatchType, patchBytes, v1.PatchOptions{})
 }
```
Thinking a little more about this. Why don't we simply replace this helper with a call to `kubevirt.Client().VirtualMachine(vmNamespace).Start()` and drop the helper?

RamLavi replied:
I agree. will remove.

0xFelix commented on tests/clone_test.go:
```diff
@@ -839,22 +694,22 @@ var _ = Describe("VirtualMachineClone Tests", Serial, func() {
 
-func stopCloneVM(virtClient kubecli.KubevirtClient, vm *virtv1.VirtualMachine) (*virtv1.VirtualMachine, error) {
-	patch, err := patch.New(patch.WithAdd("/spec/runStrategy", virtv1.RunStrategyHalted)).GeneratePayload()
+func stopCloneVM(vmName, vmNamespace string) (*virtv1.VirtualMachine, error) {
```
Same

0xFelix commented on tests/clone_test.go:
```diff
@@ -874,3 +729,149 @@ func createSourceVM(options ...libvmi.Option) (*virtv1.VirtualMachine, error) {
 	return virtClient.VirtualMachine(vm.Namespace).Create(context.Background(), vm, v1.CreateOptions{})
 }
+
+func createSnapshot(vmName, vmNamespace string) *snapshotv1.VirtualMachineSnapshot {
```
Can we reduce `createSnaphot` and `waitSnapshotReady` into a single helper? I only see them used together.

dhiller replied:
Maybe use a fluent interface here?

RamLavi replied:
I kinda prefer the "function one one thing" approach. But I'm not gonna insist on it - you call .

0xFelix commented on tests/clone_test.go:
```diff
@@ -874,3 +729,149 @@ func createSourceVM(options ...libvmi.Option) (*virtv1.VirtualMachine, error) {
+	return vmClone
+}
+
+func createCloneAndWaitForFinish(vmClone *clone.VirtualMachineClone) {
```
Pass name/namespace of the clone instead. I don't like that this is modifying the pointer that was passed in.

RamLavi replied:
>Pass name/namespace of the clone instead

We can't , the object is used to be created. 

> I don't like that this is modifying the pointer that was passed in.

That's indeed not needed. will change.

0xFelix replied:
Might make sense to return the created clone? (possibly in a follow up and only if it is ever used again)

0xFelix commented on tests/clone_test.go:
```diff
@@ -874,3 +729,149 @@ func createSourceVM(options ...libvmi.Option) (*virtv1.VirtualMachine, error) {
+	}, 3*time.Minute, 3*time.Second).Should(Equal(clone.Succeeded), "clone should finish successfully")
+}
+
+func expectVMRunnable(vm *virtv1.VirtualMachine, login console.LoginToFunction) *virtv1.VirtualMachine {
```
Same

RamLavi replied:
This function is over bloated, I plan to tackle it in a follow up PR.

RamLavi commented:
[Change](https://github.com/kubevirt/kubevirt/compare/36e297b94fb39df0fa9daca0206386d667b4785a..97367314ef7f0830911463950d454fa76d039ea6): Address @0xFelix 's review

0xFelix reviewed (commented):
/approve

0xFelix commented:
Also, looks like there are uncommitted changes!

RamLavi commented:
> Also, looks like there are uncommitted changes!

wdym?

RamLavi commented:
@0xFelix I'm going to break this PR
